Add workspaceType best practice doc (#4053)
kcp-dev/kcp: docs/content/concepts/workspaces/workspace-types-best-practices.md

---
description: >
  Guidance on how to design, deploy, and consume WorkspaceTypes safely.
---

# WorkspaceType Best Practices

`WorkspaceType` is a powerful extension point in kcp, but it is also one of

> the most frequently misused. The core reason is that a `WorkspaceType` is

Reviewer: ```suggestion
`WorkspaceType` is a powerful extension point in kcp, but it is also very
easily misused. The core reason is that a `WorkspaceType` is
```

Not saying I disagree but let's make the wording softer :D

Rule of thumb: if you are tempted to put business logic behind an
initializer, the workspace probably does not need the initializer at all.
Provision asynchronously and let the consumer observe readiness through a
status field on their own resource.

## Initialization Best Practices

Initializers turn workspace creation into a distributed rendezvous. Every
controller that owns an initializer is, for as long as it holds the
initializer, a hard dependency of `Workspace.status.phase = Ready`.

### Prefer non-blocking patterns

Most "initialization" work does not actually need to block the workspace
from becoming ready. Prefer this order of choices:

1. **No initializer at all.** Have your controller watch for new
   `LogicalCluster` objects (or workspace-scoped resources it cares about)
   and reconcile asynchronously. The workspace becomes `Ready` immediately
   and the user observes provisioning progress on your own CR's status.
2. **Initializer used only for truly blocking invariants.** Something the
   workspace is *broken* without — for example, a `ClusterRoleBinding` that
   grants the workspace owner any access at all, or an `APIBinding` the

>    workspace is contractually expected to have on day one.

Reviewer: ```suggestion
2. **Initializer used only for truly blocking invariants.** Something the
   workspace is *broken* without — for example, a `ClusterRoleBinding` that
   grants the workspace owner any access at all, or assigning a cost center
   to correctly assign costs generated from the workspace.
```

I wouldn't put an exception for APIBindings in the documentation. Sure it's possible but I wouldn't give someone the option of pointing to our documentation and saying "look, they are saying its okay when X".

3. **Avoid chaining multiple blocking initializers from different teams on
   the same type.** Each one becomes a single point of failure for every
   workspace of that type. If any controller is down or slow, every new
   workspace of that type stalls.

### When you do need an initializer

- Keep the initializer controller's reconcile loop **fast and idempotent**.
  It sits on the critical path of workspace creation; a p99 of several
  seconds multiplies across bursts.
- Make the controller **highly available**. It is infrastructure for
  every workspace of this type, not a tenant-scoped component.
- Use the
  [`initializingworkspaces` virtual workspace](./workspace-initialization.md#the-initializingworkspaces-virtual-workspace)
  and patch-based updates — do not try to `Update` `LogicalCluster` objects.
- Treat the set of initializers on a type as **append-only from the user's
  perspective**. Adding a new blocking initializer to an existing type
  changes the readiness contract for workspaces already in flight.

### Anti-patterns

> - Using initializers to run long-running bootstrap scripts.

Reviewer: ```suggestion
- Using initializers to run long-running bootstrap processes.
```

## Termination Best Practices

Terminators are the counterpart to initializers and carry the same risk
profile in reverse: every terminator is a hard dependency of the workspace
actually going away. A stuck terminator is a stuck workspace deletion.

- Prefer regular Kubernetes **finalizers on workspace-scoped resources**
  for cleanup that is internal to the workspace. Terminators should only
  run logic that must happen *outside* the workspace (for example, freeing
  an external account, releasing a provider-side record).
- Terminators run with **no guaranteed ordering** relative to each other
  or to in-workspace finalizers. Do not rely on another controller having
  already cleaned up a particular resource.
- Make the cleanup logic **idempotent and side-effect-safe on retry** —
  terminators can and will run multiple times.
- Have a clear plan for **what happens if the external system is
  unreachable**. A terminator that blocks forever on a 500 from a
  downstream API leaks workspace objects indefinitely.
- Emit a status condition the operator can see. "Stuck in terminating" is
  the single most common workspace-lifecycle support request; make it easy

>   to triage.

Reviewer 2: I am thinking these two sections should go into the respective initializers/terminators docs (e.g workspace-initialization.md).

This would make this best practices document more focused on WorkspaceTypes themselves and give people looking up initializers/terminators specifically, the required info.

Author: I intentionally left it from there, as those two docs are technical, on how to use it. And here is more of "good practices". So my idea was total oposite :D

Author: So whats the consensus? :D

Reviewer: I see Simons point. Could add a best practices section to each and link to those from here?

Author: Im not sure I like the shape of those docs in general. Having 2 docs for initialising and terminating, and now splitting docs is just creating more "clicky clicky" experience. 

Maybe we need to rewamp this section of docs in general giving its own tab for these docs? Or squasing init/terminators docs into one page.

Author: <img width="3246" height="1442" alt="image" src="https://github.com/user-attachments/assets/3aaa0ddf-0f43-4637-a356-e10c01c16989" />
Would this be better? As now its all in "one big list"

  `org-v2`) rather than mutating `org` in place, whenever the change alters
  the readiness contract.
- Use `extend.with` to share behavior between the old and new type during
  the transition, so you are not duplicating the whole definition.
- Give consumers time to migrate before retiring the old type.

### Do not encode per-workspace configuration in the type

Because the type is shared, anything that varies per workspace (quotas,
cost center, feature flags, account IDs) must live on a **provider-owned
CR** inside or next to the workspace — not on the `WorkspaceType` itself.
The type says "this is an organization workspace"; a separate
`Organization` CR says "...and specifically, it's for team X with quota Y".

## Provider Bootstrapping Without Initializers

There is a genuine chicken-and-egg problem for provider controllers:
a provider can only reconcile workspaces it can see, and for most
providers "see" means "has an `APIBinding` to the provider's API". A
freshly created workspace has no such binding yet. The naive fix — solve
it with a blocking initializer on the `WorkspaceType` — pulls every
provider onto the critical path of workspace creation, which is exactly

> what this page is trying to avoid.

Reviewer: ```suggestion
Providers can only reconciler workspaces they see through their virtual
workspaces, that means the workspaces require the APIBinding.

Solving this issue by giving each provider as a blocking initializer on
the `WorkspaceType` makes **every** provider a critical component for
workspace creation. If **one** provider runs in an issue the workspace
creation is blocked until that provider recovers.
```

A bit softer wording while keeping the core issue in view I think :D
